File: water_meter.py
# water_meter.py
import json
import os
import sys
import argparse
from datetime import datetime, timedelta

DATA_FILE = "water_data.json"
DEFAULT_UNIT = "L"

class WaterMeter:
    def __init__(self):
        self.data = []
        self.load()

    def load(self):
        if os.path.exists(DATA_FILE):
            with open(DATA_FILE, "r") as f:
                try:
                    self.data = json.load(f)
                except:
                    self.data = []

    def save(self):
        with open(DATA_FILE, "w") as f:
            json.dump(self.data, f, indent=2)

    def add(self, value, unit="L", date=None):
        if date is None:
            date = datetime.now().strftime("%Y-%m-%d")
        entry = {"date": date, "value": value, "unit": unit}
        self.data.append(entry)
        self.save()
        print(f"✅ Added: {value} {unit} on {date}")

    def list(self):
        if not self.data:
            print("No data.")
            return
        print("\n📋 All readings:")
        for e in self.data:
            print(f"  {e['date']}: {e['value']} {e['unit']}")

    def stats(self, days=7, unit=None):
        if not self.data:
            print("No data.")
            return
        cutoff = datetime.now() - timedelta(days=days)
        filtered = [e for e in self.data if datetime.strptime(e["date"], "%Y-%m-%d") >= cutoff]
        if not filtered:
            print(f"No data in the last {days} days.")
            return
        # Convert all to the requested unit
        target_unit = unit if unit else DEFAULT_UNIT
        total = 0
        for e in filtered:
            val = e["value"]
            if e["unit"] == "m3" and target_unit == "L":
                val *= 1000
            elif e["unit"] == "L" and target_unit == "m3":
                val /= 1000
            total += val
        avg = total / len(filtered)
        print(f"\n📊 Statistics (last {days} days) – unit: {target_unit}")
        print(f"  Total: {total:.2f} {target_unit}")
        print(f"  Average per day: {avg:.2f} {target_unit}")

    def reset(self):
        self.data = []
        self.save()
        print("All data cleared.")

def main():
    parser = argparse.ArgumentParser(description="Water Meter Widget")
    subparsers = parser.add_subparsers(dest="cmd", required=True)

    add_parser = subparsers.add_parser("add")
    add_parser.add_argument("value", type=float)
    add_parser.add_argument("unit", nargs="?", default="L", choices=["L", "m3"])
    add_parser.add_argument("--date", help="YYYY-MM-DD", default=None)

    stats_parser = subparsers.add_parser("stats")
    stats_parser.add_argument("--days", type=int, default=7)
    stats_parser.add_argument("--unit", choices=["L", "m3"], default=None)

    subparsers.add_parser("list")
    subparsers.add_parser("reset")

    args = parser.parse_args()
    meter = WaterMeter()

    if args.cmd == "add":
        meter.add(args.value, args.unit, args.date)
    elif args.cmd == "stats":
        meter.stats(args.days, args.unit)
    elif args.cmd == "list":
        meter.list()
    elif args.cmd == "reset":
        meter.reset()

if __name__ == "__main__":
    main()

File: water_meter.js
// water_meter.js
#!/usr/bin/env node
const fs = require('fs');
const { program } = require('commander');

const DATA_FILE = 'water_data.json';

class WaterMeter {
    constructor() {
        this.data = [];
        this.load();
    }

    load() {
        if (fs.existsSync(DATA_FILE)) {
            try {
                this.data = JSON.parse(fs.readFileSync(DATA_FILE));
            } catch (e) {
                this.data = [];
            }
        }
    }

    save() {
        fs.writeFileSync(DATA_FILE, JSON.stringify(this.data, null, 2));
    }

    add(value, unit = 'L', date = null) {
        if (!date) {
            date = new Date().toISOString().slice(0,10);
        }
        this.data.push({ date, value, unit });
        this.save();
        console.log(`✅ Added: ${value} ${unit} on ${date}`);
    }

    list() {
        if (this.data.length === 0) {
            console.log('No data.');
            return;
        }
        console.log('\n📋 All readings:');
        for (const e of this.data) {
            console.log(`  ${e.date}: ${e.value} ${e.unit}`);
        }
    }

    stats(days = 7, unit = null) {
        if (this.data.length === 0) {
            console.log('No data.');
            return;
        }
        const cutoff = new Date();
        cutoff.setDate(cutoff.getDate() - days);
        const filtered = this.data.filter(e => new Date(e.date) >= cutoff);
        if (filtered.length === 0) {
            console.log(`No data in the last ${days} days.`);
            return;
        }
        const targetUnit = unit || 'L';
        let total = 0;
        for (const e of filtered) {
            let val = e.value;
            if (e.unit === 'm3' && targetUnit === 'L') val *= 1000;
            else if (e.unit === 'L' && targetUnit === 'm3') val /= 1000;
            total += val;
        }
        const avg = total / filtered.length;
        console.log(`\n📊 Statistics (last ${days} days) – unit: ${targetUnit}`);
        console.log(`  Total: ${total.toFixed(2)} ${targetUnit}`);
        console.log(`  Average per day: ${avg.toFixed(2)} ${targetUnit}`);
    }

    reset() {
        this.data = [];
        this.save();
        console.log('All data cleared.');
    }
}

program
    .command('add <value>')
    .option('--unit <unit>', 'L or m3', 'L')
    .option('--date <date>', 'YYYY-MM-DD')
    .action((value, options) => {
        const meter = new WaterMeter();
        meter.add(parseFloat(value), options.unit, options.date);
    });

program
    .command('list')
    .action(() => {
        const meter = new WaterMeter();
        meter.list();
    });

program
    .command('stats')
    .option('--days <days>', 'Number of days', parseInt, 7)
    .option('--unit <unit>', 'L or m3')
    .action((options) => {
        const meter = new WaterMeter();
        meter.stats(options.days, options.unit);
    });

program
    .command('reset')
    .action(() => {
        const meter = new WaterMeter();
        meter.reset();
    });

program.parse(process.argv);
